Support Instead of Trigger On Views

deepakshi-mittal opened the pull request:
### Description
Currently Postgres does not support Instead of Trigger On Views for DML queries.

On running such a query on VIEW, query rewriting substitutes view with underlying base table.
Added a hook for skipping the substitution for DML query on the VIEW.

This change will add support for Instead of Triggers for DML queries when using Babelfish.

Task: BABEL-2170/BABEL-4532/BABEL-4514

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/272

### Issues Resolved
 BABEL-2170/BABEL-4532/BABEL-4514

### Test Scenarios Covered
Instead of Trigger (IOT) on View for Insert/Update/Delete queries
Transition tables inside IOT on View
IOT calling itself. IOT A <-> IOT A
IOT A <-> IOT B 
IOT with transition tables
IOT Cross database
IOT Cross Schema
IOT for Transaction, Commit, Rollback, Rollback to Savepoint
IOT for Bulk insert
Same name trigger not allowed for View/Table
DROP Trigger cross Schema

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deepakshi-mittal committed:
Support Instead of Trigger On Views

Currently Postgres does not support Instead of Trigger On Views for DML
queries.

On running such a query on VIEW, query rewriting substitutes view with
underlying base table.
Added a hook for skipping the substitution for DML query on the VIEW.

This change will add support for Instead of Triggers for DML queries
when using Babelfish.

Task: BABEL-2170/BABEL-4532/BABEL-4514

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Added test cases from previously failing scripts

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Merged and removed extra hook and renamed oid variable

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Added test to all upgrade schedule files

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

xhfanhe commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -719,6 +725,61 @@ plsql_TriggerRecursiveCheck(ResultRelInfo *resultRelInfo)
+				return false; /** A trigger called recursively by itself*/
+			}
+			else if (list_member_oid(triggerInvocationSequence, current_tgoid))
+            {
```
Please fix the indentation here

xhfanhe commented on test/JDBC/expected/BABEL-2170-vu-prepare.out:
```diff
@@ -0,0 +1,105 @@
+-- We will create two databases db1_BABEL2170, db2_BABEL2170
```
How come this get passed in the JDBC test? I thought JDBC is single-db mode?

deepakshi-mittal replied:
I used SELECT set_config('babelfishpg_tsql.migration_mode', 'multi-db', false); 
Its used in many places in JDBC test.  Some Example file:
babelfish_extensions/test/JDBC/input/BABEL-1446.sql
babelfish_extensions/test/JDBC/input/BABEL-1435.sql

deepakshi-mittal replied:
Ignored file containing multi-db test from jdbc schedule.
Removed set config change for multi-db
Added basic test cases seperately to verify IO Triggers.

xhfanhe commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -719,6 +725,61 @@ plsql_TriggerRecursiveCheck(ResultRelInfo *resultRelInfo)
+                /** A trigger called recursively by another trigger */
+                ereport(ERROR,
+                        (errcode(ERRCODE_SYNTAX_ERROR),
+                         errmsg("Maximum stored procedure, function, trigger, or view nesting level exceeded (limit 32)")));
```
Do we really checked 32? We should modify the error message if it's not the real case.

deepakshi-mittal replied:
The error message is fixed in cases even if recursion level is not 32.
 I also added a check if it exceeds 32 depth

if (list_length(triggerInvocationSequence) > 32)
	ereport(ERROR,(errcode(ERRCODE_SYNTAX_ERROR), errmsg("Maximum stored procedure, function, trigger, or view nesting level exceeded (limit 32)")));

deepakshi-mittal committed:
Added condition for recursive trigger reaching limit 32

Fixed indentation

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Addressed Code review suggestion for muti-db test cases

Ignored file containing multi-db test from jdbc schedule.
Removed set config change for multi-db
Added basic test cases seperately to verify IO Triggers.

Task:BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

xhfanhe commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -719,6 +725,67 @@ plsql_TriggerRecursiveCheck(ResultRelInfo *resultRelInfo)
+				/** A trigger called recursively by another trigger */
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("Maximum stored procedure, function, trigger, or view nesting level exceeded (limit 32)")));
```
This can be combined with previous if check. And this is not a syntax error, ERRCODE_PROGRAM_LIMIT_EXCEEDED may be a better error code.
The comment is also a bit confusing here - I suggest change it into something like "loop trigger call".

deepakshi-mittal committed:
Merged if condition into one, defined TRIGGER_MAX_NEST_LEVEL

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -719,6 +725,61 @@ plsql_TriggerRecursiveCheck(ResultRelInfo *resultRelInfo)
+			{
+				return false; /** Loop trigger call by itself*/
+			}
+			else if (list_length(triggerInvocationSequence) > 32 || list_member_oid(triggerInvocationSequence, current_tgoid))
```
Where did you get this number ? 
Can this number be adjusted in Sql server ?

deepakshi-mittal replied:
https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/triggers/create-nested-triggers.md
It says DML and DDL triggers can be nested up to 32 levels. I dont think this number(level 32) can be changed. It can only change whether to allow triggers of that type or not.  
Also sp_configure and alter database is not yet supported in Babelfish, through which settings related to nesting and recursion can be changed.

Above document has links for same
To set the nested triggers option
[sp_configure (Transact-SQL)](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/system-stored-procedures/sp-configure-transact-sql.md)
To set the RECURSIVE_TRIGGERS database option

[ALTER DATABASE SET Options (Transact-SQL)](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/statements/alter-database-transact-sql-set-options.md)

forestkeeper commented on test/JDBC/jdbc_schedule:
```diff
@@ -128,6 +128,9 @@ ignore#!#case_insensitive_collation-before-13-5-vu-cleanup
 ignore#!#jira-BABEL-3504-upgrade-vu-prepare
 ignore#!#jira-BABEL-3504-upgrade-vu-verify
 ignore#!#jira-BABEL-3504-upgrade-vu-cleanup
+ignore#!#BABEL-2170-vu-prepare
```
Why it needs to ignore in jdbc schedule test ?

deepakshi-mittal replied:
Xiaohui reviewed it last week and had some concerns about multi-db tests running in jdbcs tests. There was one test case in BABEL-2170-vu-verify which needs to run in multi-db setting. 
All the tests will be executed in upgrade tests now in multi-db mode. Since JDBC tests are in single-db, I added ignore here.
